    Week6-Homework

    Q11. Learn about the affix tagger (type help(nltk.AffixTagger)). Train an affix tagger and run it on some new text. Experiment with different settings for the affix length and the minimum word length. Discuss your findings.

In [1]:
import nltk
from nltk.corpus import brown

#Loading the Brown corpus
corpus = brown.tagged_sents(categories='news')

#Splitting the data into training and testing sets
train_data = corpus[:int(0.8 * len(corpus))]
test_data = corpus[int(0.8 * len(corpus)):]

#Training an AffixTagger with different settings
def train_affix_tagger(affix_length, min_word_length):
    affix_tagger = nltk.AffixTagger(train_data, affix_length=affix_length, min_stem_length=min_word_length)
    return affix_tagger

#Function to evaluate the tagger
def evaluate_tagger(tagger, test_data):
    return tagger.evaluate(test_data)

#Trying different settings
affix_lengths = [2, 3, 4]  # Try different lengths for prefixes and suffixes
min_word_lengths = [2, 3, 4]  # Try different minimum word lengths

for affix_length in affix_lengths:
    for min_word_length in min_word_lengths:
        print(f"Affix length: {affix_length}, Min word length: {min_word_length}")
        affix_tagger = train_affix_tagger(affix_length, min_word_length)
        accuracy = evaluate_tagger(affix_tagger, test_data)
        print(f"Accuracy: {accuracy}")


Affix length: 2, Min word length: 2


/var/folders/32/d10tw0x10ngd3fsg70yr4v9m0000gn/T/ipykernel_37555/4138137748.py:18: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  return tagger.evaluate(test_data)


Accuracy: 0.2050195096102895
Affix length: 2, Min word length: 3
Accuracy: 0.14697239751433114
Affix length: 2, Min word length: 4
Accuracy: 0.11012091141191772
Affix length: 3, Min word length: 2
Accuracy: 0.17780239895948746
Affix length: 3, Min word length: 3
Accuracy: 0.1267402090659473
Affix length: 3, Min word length: 4
Accuracy: 0.09470591068933956
Affix length: 4, Min word length: 2
Accuracy: 0.13290620935497857
Affix length: 4, Min word length: 3
Accuracy: 0.09865600462450022
Affix length: 4, Min word length: 4
Accuracy: 0.061660002890312635


Here are some observations based on the results:

Affix Length:
Generally, shorter affix lengths (2 or 3) tend to have higher accuracies compared to longer ones (4). This suggests that shorter affixes might capture more meaningful patterns in the data.
The accuracy decreases as the affix length increases, which might indicate that longer affixes result in more sparse and less reliable patterns.

Minimum Word Length:
Similar to affix length, shorter minimum word lengths (2 or 3) tend to yield higher accuracies compared to longer ones (4).
However, the difference in accuracy between different minimum word lengths is less pronounced compared to affix length. This could suggest that the impact of minimum word length on accuracy is not as significant as affix length.

Overall:
The highest accuracy achieved is around 20.5%, which indicates that the AffixTagger alone might not be sufficient for achieving high accuracy in part-of-speech tagging tasks.
It's important to note that these results are specific to the Brown corpus and may vary with different datasets.

    Q14. Use sorted() and set() to get a sorted list of tags used in the Brown Corpus, removing duplicates.

In [2]:
import nltk
from nltk.corpus import brown
nltk.download('universal_tagset')

#Getting the tags from the Brown Corpus
tags = brown.tagged_words(tagset='universal')

#Extracting unique tags and sort them
unique_tags = sorted(set(tag for word, tag in tags))

print(unique_tags)


[nltk_data] Downloading package universal_tagset to
[nltk_data]     /Users/db/nltk_data...
[nltk_data]   Package universal_tagset is already up-to-date!


['.', 'ADJ', 'ADP', 'ADV', 'CONJ', 'DET', 'NOUN', 'NUM', 'PRON', 'PRT', 'VERB', 'X']


    Q15. Write programs to process the Brown Corpus and find answers to the following questions:

    a. Which nouns are more common in their plural form, rather than their singular form? (Only consider regular plurals, formed with the -s suffix.)

In [3]:
import nltk
from nltk.corpus import brown
nltk.download('averaged_perceptron_tagger')

# Load the Brown Corpus and tokenize words
words = brown.words()
tagged_words = nltk.pos_tag(words)

# Initialize set to store singular nouns
singular_nouns = set()

# Iterate through tagged words to identify singular nouns
for word, tag in tagged_words:
    if tag == 'NN':  # Singular noun
        singular_nouns.add(word)

# Initialize dictionary to store counts of plural nouns
plural_counts = {}

# Iterate through tagged words again to count plural nouns
for word, tag in tagged_words:
    if tag == 'NNS':  # Plural noun
        if word in singular_nouns:  # Check if the singular form exists
            plural_counts[word] = plural_counts.get(word, 0) + 1

# Identify nouns that are more common in their plural form than singular form
more_common_plural = [word for word, count in plural_counts.items() if count > 0]

# Print the list of nouns
print("Nouns more common in their plural form than singular form:")
for word in more_common_plural:
    print(word)


[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/db/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


Nouns more common in their plural form than singular form:
personnel
criticisms
employes
means
reservoirs
kept
police
drafts
we'll
years'
parks
ours
beliefs
clubs
blows
breaks
crowds
feels
boys
reflexes
guys
works
strokes
triumphs
People
bombs
stairs
pardon
east
troops
data
employ
grandchildren
eggs
ribs
months'
m.p.h.
headquarters
roles
robes
yarn
lamps
legs
relationships
dreams
toys
films
sofas
armchairs
tours
downstairs
propaganda
mechanisms
brethren
women's
bags
drops
teeth
bats
tribes
guns
shadows
lags
seams
valleys
magic
hooks
isn't
stems
hopefuls
hysteria
reads
throats
baths
won't
they'd
thru
cafes
thugs
sixty
rags
clouds
cannot
nightclubs
orchestras
proud
formulas
compete
orchestra
cooks
discs
left
didn't
frank
souls
epigrams
limbs
sonatas
wet
chantey
contours
suburbs
glad
lobes
salesmen
keys
lawns
maze
marinas
lbs.
milligrams
menu
evil
sphere
mankind
maps
peas
thousand
bonzes
thou
yours
mobs
gonna
doesn't
You'll
lengths
thighs
mulch
galleys
diagrams
drill
puzzle
kid
herbs
visi

    b. Which word has the greatest number of distinct tags? What are they, and what do they represent?

In [6]:
import nltk
from nltk.corpus import brown
nltk.download('brown')
nltk.download('tagsets')

# Load the Brown Corpus and tokenize words
tagged_words = nltk.corpus.brown.tagged_words()

# Initialize a dictionary to store distinct tags for each word
word_tags = {}

# Iterate through tagged words to count distinct tags for each word
for word, tag in tagged_words:
    if word not in word_tags:
        word_tags[word] = set()  # Initialize set for new word
    word_tags[word].add(tag)  # Add tag to the set for the word

# Find the word with the greatest number of distinct tags
max_word = max(word_tags, key=lambda word: len(word_tags[word]))
max_tags = word_tags[max_word]

# Print the word and its distinct tags
print(f"The word with the greatest number of distinct tags is '{max_word}'")
print("Distinct tags and their meanings:")
for tag in max_tags:
    print(f"{tag}: {nltk.help.upenn_tagset(tag)}")


[nltk_data] Downloading package brown to /Users/db/nltk_data...
[nltk_data]   Package brown is already up-to-date!
[nltk_data] Downloading package tagsets to /Users/db/nltk_data...
[nltk_data]   Unzipping help/tagsets.zip.


The word with the greatest number of distinct tags is 'that'
Distinct tags and their meanings:
No matching tags found.
WPO-NC: None
No matching tags found.
CS: None
DT: determiner
    all an another any both del each either every half la many much nary
    neither no some such that the them these this those
DT: None
No matching tags found.
CS-HL: None
No matching tags found.
DT-NC: None
No matching tags found.
CS-NC: None
No matching tags found.
NIL: None
No matching tags found.
WPS-NC: None
No matching tags found.
WPO: None
No matching tags found.
QL: None
No matching tags found.
WPS: None
No matching tags found.
WPS-HL: None


    c.List tags in order of decreasing frequency. What do the 20 most frequent tags represent?

In [7]:
import nltk
from nltk.corpus import brown
from collections import Counter

# Load the Brown Corpus and tokenize words
tagged_words = nltk.corpus.brown.tagged_words()

# Count the occurrences of each tag
tag_counts = Counter(tag for word, tag in tagged_words)

# Sort the tags by frequency in descending order
sorted_tags = sorted(tag_counts.items(), key=lambda x: x[1], reverse=True)

# Print the 20 most frequent tags
print("The 20 most frequent tags and their frequencies:")
for tag, count in sorted_tags[:20]:
    print(f"{tag}: {count}")


The 20 most frequent tags and their frequencies:
NN: 152470
IN: 120557
AT: 97959
JJ: 64028
.: 60638
,: 58156
NNS: 55110
CC: 37718
RB: 36464
NP: 34476
VB: 33693
VBN: 29186
VBD: 26167
CS: 22143
PPS: 18253
VBG: 17893
PP$: 16872
TO: 14918
PPSS: 13802
CD: 13510


    d. Which tags are nouns most commonly found after? What do these tags represent?

In [8]:
import nltk
from nltk.corpus import brown
from collections import Counter

#Loading the Brown Corpus and tokenize words
tagged_words = nltk.corpus.brown.tagged_words()

#Initializing a dictionary to store tags that follow nouns and their frequencies
tags_after_nouns = {}

#Iteratimng through tagged words to identify tags that follow nouns and count their occurrences
for i in range(len(tagged_words) - 1):
    current_word, current_tag = tagged_words[i]
    next_word, next_tag = tagged_words[i + 1]
    if current_tag.startswith('NN'):  # Check if the current word is a noun
        if current_tag not in tags_after_nouns:
            tags_after_nouns[current_tag] = Counter()  # Initialize counter for the tag
        tags_after_nouns[current_tag][next_tag] += 1  # Increment the count for the following tag

#Finding the most common tags that follow each type of noun
most_common_tags_after_nouns = {noun_tag: tag_counter.most_common(1) for noun_tag, tag_counter in tags_after_nouns.items()}

#Printing the most common tags that follow each type of noun
print("Most common tags that follow each type of noun:")
for noun_tag, common_tags in most_common_tags_after_nouns.items():
    print(f"{noun_tag}: {common_tags[0][0]} ({common_tags[0][1]} occurrences)")


Most common tags that follow each type of noun:
NN-TL: NN-TL (1914 occurrences)
NN: IN (42256 occurrences)
NNS: IN (14505 occurrences)
NN-HL: IN-HL (240 occurrences)
NN$-TL: NN (136 occurrences)
NN$: NN (887 occurrences)
NNS-HL: IN-HL (110 occurrences)
NNS-TL: , (381 occurrences)
NNS$: NN (116 occurrences)
NNS$-TL: NN-TL (38 occurrences)
NN-TL-HL: NN-TL-HL (24 occurrences)
NNS-TL-HL: :-HL (6 occurrences)
NN$-HL: NN-HL (19 occurrences)
NNS$-HL: NNS-HL (3 occurrences)
NN-NC: '' (35 occurrences)
NN+HVZ: VBN (3 occurrences)
NN+BEZ: VBN (6 occurrences)
NNS-NC: BER-NC (3 occurrences)
NN+MD: VB (2 occurrences)
NNS-TL-NC: '' (2 occurrences)
NN-TL-NC: '' (3 occurrences)
NNS$-TL-HL: NN-TL-HL (1 occurrences)
NNS$-NC: JJ (1 occurrences)
NN+NN-NC: , (1 occurrences)
NN+IN: JJ (1 occurrences)
NN+HVD-TL: BEN (1 occurrences)
NN+BEZ-TL: VBN (1 occurrences)
NN+HVZ-TL: VBN (1 occurrences)
NNS+MD: ABN (1 occurrences)
